test/core/xds/xds_client_test.cc

```diff
@@ -3779,6 +3804,237 @@ TEST_F(XdsClientTest, AdsReadWaitsForHandleRelease) {
   EXPECT_TRUE(stream->Orphaned());
 }
 
+TEST_F(XdsClientTest, FallbackAndRecover) {
+  FakeXdsBootstrap::FakeXdsServer primary_server(kDefaultXdsServerUrl);
+  FakeXdsBootstrap::FakeXdsServer fallback_server("fallback_xds_server");
+  // Regular operation
+  InitXdsClient(FakeXdsBootstrap::Builder()
+                    .AddServer(primary_server)
+                    .AddServer(fallback_server));
+  // Start a watch for "foo1".
+  auto watcher = StartFooWatch("foo1");
+  EXPECT_THAT(GetServerConnections(), ::testing::ElementsAre(::testing::Pair(
+                                          kDefaultXdsServerUrl, true)));
+  EXPECT_THAT(GetResourceCounts(),
+              ::testing::ElementsAre(::testing::Pair(
+                  ResourceCountLabelsEq(XdsClient::kOldStyleAuthority,
+                                        XdsFooResourceType::Get()->type_url(),
+                                        "requested"),
+                  1)));
+  // XdsClient should have created an ADS stream.
+  auto stream = WaitForAdsStream();
+  ASSERT_TRUE(stream != nullptr);
+  // XdsClient should have sent a subscription request on the ADS stream.
+  auto request = WaitForRequest(stream.get());
+  ASSERT_TRUE(request.has_value());
+  // Send a response.
+  stream->SendMessageToClient(
+      ResponseBuilder(XdsFooResourceType::Get()->type_url())
+          .set_version_info("1")
+          .set_nonce("A")
+          .AddFooResource(XdsFooResource("foo1", 6))
+          .Serialize());
+  // XdsClient should have delivered the response to the watcher.
+  auto resource = watcher->WaitForNextResource();
+  ASSERT_NE(resource, nullptr);
+  EXPECT_EQ(resource->name, "foo1");
+  EXPECT_EQ(resource->value, 6);
+  TriggerConnectionFailure(primary_server,
```

Comment:
> I tried my best to implement your suggestions, ptal.

This is a lot better, but the order of steps still isn't quite right.  I think there are two principles here:
- For each input we trigger on the client, we should check all of the results of that input before we provide the next input.
- When checking the results of a given input, we should always check the local results (i.e., watcher events and metric data) before we remote results (i.e., requests sent to the server), because we shouldn't have to wait for remote events to complete (which may involve waiting for activity in other threads) before seeing the local results.

Given that, I think the order we want is:

1. Input: Get initial response from primary server.
   1. Result (local): Resource is delivered to watcher.
   2. Result (local): Metrics show 1 resource update and 1 cached resource.
   3. Result (remote): Client sends ACK to server.
2. Input: Trigger connection failure to primary.
   1. Result (local): The error is reported to the watcher.
   2. Result (local): The metrics show the channel as being unhealthy.
3. Input: Trigger stream failure.
   1. Result (local): The metrics still show the channel as being unhealthy.
   2. Result (remote): The client starts a new stream and sends a subscription message.
4. Input: Start second watch for foo1 (already cached).
   1. Result (local): New watcher gets the cached resource.
   2. Result (local): New watcher gets the error from the channel state.
5. Input: Start watch for foo2 (not already cached).
    1. Result (local): Metrics show a healthy channel to the fallback server.
    2. Result (remote): Client sent a new request for both resources on the stream to the primary.
    3. Result (remote): Client created a stream to the fallback server and sent a request on that stream for both resources.
6. Input: Fallback server sends a response with both resources.
   1. Result (local): Watcher for foo2 gets resource.
   2. Result (local): Metrics show an update from fallback server.
   3. Result (remote): Client sends ACK to fallback server.
7. Input: Primary server sends a response containing both resources.
   1. Result (local): Metrics show that we've closed the channel to the fallback server and received resource updates from the primary server.
   2. Result (remote): The stream to the fallback server has been orphaned.

Please add a one-line comment before each input and each result check, so that it's clear what's happening.

> Note that no error is returned for the original watcher here:

That seems like a bug.  We're not triggering fallback here, so we should report the error to the watcher.